nondex-common/src/main/java/edu/illinois/nondex/common/SystematicRandom.java

```diff
@@ -0,0 +1,142 @@
+*/
+
+package edu.illinois.nondex.common;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.Stack;
+
+public class SystematicRandom extends Random {
+    public static int STARTING_COUNT = 59;
+    int replayIndex;
+    int count;
+    Stack<StackElement> choice;
+    String logFileName;
+    Charset utf8 = StandardCharsets.UTF_8;
+    List<String> lines;
+
+    public SystematicRandom() {
+        logFileName = System.getenv("logFileName");
+        if (logFileName != null) {
+            File file = new File(this.logFileName);
+            if (!file.exists()) {
+                choice = new Stack<StackElement>();
+                replayIndex = 0;
+                count = 0;
+            } else {
+                choice = new Stack<StackElement>();
+                try {
+                    lines = Files.readAllLines(Paths.get(logFileName));
+                } catch (IOException ioe) {
+                    ioe.printStackTrace();
+                }
+                Object[] choiceValues = new Object[3];
+                for (String element: lines) {
+                    String delimit = "[ ]+";
+                    if (!element.isEmpty()) {
+                        choiceValues = element.split(delimit);
+                        int last = Integer.parseInt(choiceValues[0].toString());
+                        int max = Integer.parseInt(choiceValues[1].toString());
+                        boolean explore = Boolean.parseBoolean(choiceValues[2].toString());
+                        StackElement lastMax = new StackElement( last, max, explore );
+                        choice.push(lastMax);
+                    }
+                }
+            }
+        } else {
+            System.err.println("The env variable does not exist. File could not be created");
+            System.exit(1);
+        }
+    }
+
+    public int nextInt(int max) {
+        int last;
+        count++;
+        boolean explore;
+        if (replayIndex < choice.size()) {
+            last =  choice.get(replayIndex).getLast();
+        } else {
+            last = 0;
+            if (count > STARTING_COUNT + 1) {
+                explore = true;
+            } else {
+                explore = false;
+            }
+            StackElement choiceNums = new StackElement(last, max, explore);
+            choice.push(choiceNums);
+        }
+        replayIndex++;
+        return last;
+    }
+
+    public void endRun() throws IOException {
+        while (!choice.isEmpty()) {
+            ArrayList<Object> ex = new ArrayList<>();
+            StackElement lastMax = choice.pop();
+            int last = lastMax.getLast();
+            int max = lastMax.getMax();
+            boolean explore = false;
```

Comment:
the explore here is used to flag whether the new element (last/max combination) we push to choice should be explored.  If you look inside the if branch you will see that we are changing last as well by incrementing it by one.  So I don't necessarily want the explore stored in the stack for the popped element but instead I want explore to be false unless we are pushing an element that we want to explore.